// @gka/svelteplot/dist/helpers/typeChecks.js
import { isSymbol } from './symbols.js';
import { color } from 'd3-color';
import { CSS_COLOR, CSS_COLOR_MIX, CSS_COLOR_CONTRAST, CSS_VAR, CSS_URL, CSS_RGBA } from '../constants.js';
export function isBooleanOrNull(v) {
    return v == null || typeof v === 'boolean';
}
export function isDate(v) {
    return v instanceof Date && !isNaN(v.getTime());
}
export function isDateOrNull(v) {
    return v == null || isDate(v);
}
export function isNumberOrNull(v) {
    return v == null || Number.isFinite(v) || (typeof v === 'string' && !Number.isNaN(+v));
}
export function isNumberOrNullOrNaN(v) {
    return (v == null ||
        Number.isFinite(v) ||
        Number.isNaN(v) ||
        (typeof v === 'string' && !Number.isNaN(+v)));
}
export function isStringOrNull(v) {
    return v == null || typeof v === 'string';
}
export function isSymbolOrNull(v) {
    return v == null || ((typeof v === 'string' || typeof v === 'object') && isSymbol(v));
}
export function isColorOrNull(v) {
    return (v == null ||
        (typeof v === 'string' &&
            (v === 'currentColor' ||
                CSS_VAR.test(v) ||
                CSS_COLOR.test(v) ||
                CSS_COLOR_MIX.test(v) ||
                CSS_COLOR_CONTRAST.test(v) ||
                CSS_RGBA.test(v) ||
                CSS_URL.test(v) ||
                color(v) !== null)));
}
export function isOpacityOrNull(v) {
    return v == null || (typeof v === 'number' && Number.isFinite(v) && v >= 0 && v <= 1);
}
